---
int32 error
int32 num_devices